File: python-algorithm/leetcode/problem_974.py
"""974. Subarray Sums Divisible by K
https://leetcode.com/problems/subarray-sums-divisible-by-k/
"""
import collections
from typing import List


class Solution:
    def subarrays_div_by_k(self, nums: List[int], k: int) -> int:
        ans = 0
        store = collections.Counter()
        store[0] = 1
        cur = 0
        for num in nums:
            cur = (cur + num) % k
            if cur in store:
                ans += store[cur]
            store[cur] += 1
        return ans

File: java-algo/src/main/java/com/leetcode/Problem974.java
package com.leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * 974. Subarray Sums Divisible by K
 * https://leetcode.com/problems/subarray-sums-divisible-by-k/
 */
public class Problem974 {
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer, Integer> seen = new HashMap<>();
        seen.put(0, 1);
        int sum = 0, ans = 0;
        for (int num : nums) {
            sum = ((sum + num) % k + k) % k;
            ans += seen.getOrDefault(sum, 0);
            seen.put(sum, seen.getOrDefault(sum, 0) + 1);
        }
        return ans;
    }
}
